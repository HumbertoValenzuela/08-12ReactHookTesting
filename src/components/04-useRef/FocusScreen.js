import React, { useRef } from 'react'

export const FocusScreen = () => {

    // useRef es para mantener una referencia mutable. Es posible cambiar a lo que apunta este ref sin necesidad de disparar una renderizaciÃ³n en la parte de React
    // Hook
    const inputRef = useRef();
    // console.log(inputRef); // current: input#focus.form-control

    // Focus en el input nombre .focus .select
    const handleClick = () => {
        // document.querySelector('#focus').select();

        // Usando el useRef
        inputRef.current.select();
        // console.log(inputRef);// muestra toda la referencia del input
    };   

   return (
       <div>
           <h4>Focus Screen</h4>
           <input 
            ref={ inputRef }
            // id="focus" //dejamos de ocupar, debido a useRef
            className="form-control" 
            placeholder="Su nommbre" />

           <button className="btn btn-outline-primary mt-4" onClick={ handleClick } >Focus</button>
       </div>
   )
}
