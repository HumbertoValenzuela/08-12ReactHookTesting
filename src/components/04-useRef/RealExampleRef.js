import React, { useState } from 'react';
import { MultipleCustomHokk } from '../03-examples/MultipleCustomHokk';

// Renderizar un componente de manera condicional.
// Se ocupa MultipleCustomHokk y useFetch. En el useFetch - useEffect - Existe un problema: cuando se carga la info, es posible
// que cierre la ventana antes de terminar o que tarde mucho en responder la API y la persona cierra el componente. Luego de cerrar
// se llamará al setState. Esto lanzará un ERROr. 

// Replicando el error.
// Ir useFetch y agregar un setTimeOut. Ir al sistema - botón hide/show - Muestra la api y el botón sgte quote. Mostrará el Loading de 4seg
// antes de terminar, presiona el boton show/hide. ¿Que hicimos? desmontamos un componente y despues hacemos el setState, lanza un error.
// Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

// Y la Solución. Ir useFetch - para ver explicación

export const RealExampleRef = () => {

    // State para mostrar y ocultar condicionalmente el componente MultipleCustomHokk
    const [show, setShow] = useState(false);

    return (
        <div>
            <h4>RealExampleRef</h4>
            { show && <MultipleCustomHokk />}

            <button 
            className="btn btn-primary mt-5"
            onClick={ () => { setShow( !show ); } }
            >Show/Hide
            </button>
        </div>
    )
}
