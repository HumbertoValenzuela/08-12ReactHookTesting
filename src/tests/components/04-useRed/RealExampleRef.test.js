import React from 'react';
import { shallow } from 'enzyme';
import { RealExampleRef } from '../../../components/04-useRef/RealExampleRef';

describe('Pruebas en <RealExampleRef />', () => {
    // <RealExampleRef argumento= proveer/>
    const wrapper = shallow(<RealExampleRef />);

    test('debe mostrar correctamente', () => {
        expect(wrapper).toMatchSnapshot();
        expect( wrapper.find('MultipleCustomHokk').exists() ).toBe(false);
    });

    test('debe de mostrar el componente <MultipleCustomHokk />', () => {
        // Simular Click
        wrapper.find('button').simulate('click');
        expect( wrapper.find('MultipleCustomHokk').exists() ).toBe(true);
    });
    
    
})
