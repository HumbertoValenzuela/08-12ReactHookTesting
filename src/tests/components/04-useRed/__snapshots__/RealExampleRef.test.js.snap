// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RealExampleRef /> debe mostrar correctamente 1`] = `
<div>
  <h4>
    RealExampleRef
  </h4>
  <button
    className="btn btn-primary mt-5"
    onClick={[Function]}
  >
    Show/Hide
  </button>
</div>
`;
